dspace-api/src/main/java/org/dspace/xmlworkflow/cristin/CristinFileManager.java

```diff
@@ -0,0 +1,620 @@
+        }
+
+        return aggregatedResources;
+    }
+
+    /**
+     * is the given element from an ORE description of a DSpace items a metadata bitstream?
+     *
+     * @param desc
+     * @return
+     */
+    public boolean isMetadataBitstream(Element desc) {
+        // https://w3utv-dspace01.uio.no/dspace/xmlui/bitstream/handle/123456789/982/cristin-12087.xml?sequence=2
+
+        Attribute about = desc.getAttribute("about", RDF_NS);
+        String url = about.getValue();
+
+        // FIXME: yeah yeah, this would look better with a regex
+        String[] bits = url.split("\\?");
+        String[] urlParts = bits[0].split("/");
+        String filename = urlParts[urlParts.length - 1];
+
+        if (filename.startsWith("cristin-") && filename.endsWith(".xml")) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Get an input stream which will allow us to read the bits from this remote URL
+     *
+     * @param href
+     * @return
+     * @throws MalformedURLException
+     * @throws IOException
+     */
+    public InputStream getInputStream(String href)
+            throws MalformedURLException, IOException {
+        // ingest the bitstream from the remote url
+        URL ARurl = null;
+        InputStream in = null;
+        if (href != null) {
+            try {
+                // Make sure the url string escapes all the oddball characters
+                String processedURL = encodeForURL(href);
+                // Generate a requeset for the aggregated resource
+                ARurl = new URL(processedURL);
+                in = ARurl.openStream();
+            } catch (FileNotFoundException fe) {
+                log.error("The provided URI failed to return a resource: " + href);
+            } catch (ConnectException fe) {
+                log.error("The provided URI was invalid: " + href);
+            }
+        } else {
+            throw new IOException("Could not obtain resource from " + href);
+        }
+        return in;
+    }
+
+    /**
+     * Ingest the bitstream from the given remote url with the given properties
+     *
+     * @param context
+     * @param href         Remote URL from which to retrieve the bitstream
+     * @param bsName       Name to give the bitstream in DSpace
+     * @param mimeString   Mimetype to give the bitstream in DSpace
+     * @param targetBundle Target bundle to store the bitstream in
+     * @return
+     * @throws MalformedURLException
+     * @throws IOException
+     * @throws CrosswalkException
+     * @throws AuthorizeException
+     * @throws SQLException
+     */
+    public Bitstream ingestBitstream(Context context, String href, String bsName, String mimeString, Bundle targetBundle)
+            throws MalformedURLException, IOException, CrosswalkException, AuthorizeException, SQLException {
+        Bitstream bitstream;
+
+        InputStream in = this.getInputStream(href);
```

Comment:
**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *in*. Currently, there are execution paths that do not contain closure statements, e.g., when exception is thrown by BitstreamService.create. Either a) close *in* in a try-finally block or b) close the resource by declaring *in* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).